lib/simulator/Simulator.js

```diff
@@ -216,10 +217,22 @@ export default function Simulator(injector, eventBus, elementRegistry) {
 
     const subscriptions = scope.subscriptions;
 
-    const matchingSubscriptions = filterSet(
+    let matchingSubscriptions = filterSet(
       subscriptions, subscription => eventsMatch(event, subscription.event)
     );
 
+    if (event.type === 'error' || event.type === 'escalation') {
+      const referenceSubscriptions = filterSet(
+        matchingSubscriptions, subscription => refsMatch(event, subscription.event)
+      );
+
+      if (matchingSubscriptions.every(subscription => subscription.event.boundary)
```

Comment:
I have a hard time to understand what this does. Do you remember @sombrek? Why is the boundary event check necessary, which scenarios do we cover here?

Reply:
@nikku Yes, I remember.

There are now two types of handlers: Catch-all handlers and specific handlers that have an error/escalation reference.

Scenario: When an error/escalation occurs, you have to find the nearest matching event handler. Handlers are checked in this order: specific event sub process, catch-all event sub process, specific boundary event, catch-all boundary event. If there's no match, repeat on parent scope (at least for escalation).

https://github.com/bpmn-io/bpmn-js-token-simulation/blob/7c7415b571aa5a6b602cf0f0738df8a83a253a57/lib/simulator/Simulator.js#L225-L233

The check determines whether there's a specific handler: Starting with the second part. If there's a specific event sub process, use that. Otherwise, if there are only boundary events and at least one of them has a specific reference, use that. If that check fails, there are only catch-all handlers.

Reply:
Thanks for the details. I think I'll need to further debug this "in action" to see how it works, and leave a comment for future reference.